cmd/geth/import_from_firehose.go

```diff
@@ -0,0 +1,306 @@
+		seq  uint64
+		resp *pbfirehose.Response
+	}
+	type seqBlock struct {
+		seq   uint64
+		block *types.Block
+	}
+
+	rawCh := make(chan seqResponse, bufferSize)
+	blockCh := make(chan seqBlock, bufferSize)
+	errCh := make(chan error, 1)
+	doneCh := make(chan struct{})
+
+	// Stage 1: Stream reader goroutine with sequence numbers
+	go func() {
+		defer close(rawCh)
+		var seq uint64
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				errCh <- fmt.Errorf("error receiving from stream: %w", err)
+				return
+			}
+			rawCh <- seqResponse{seq: seq, resp: resp}
+			seq++
+		}
+	}()
+
+	// Stage 2: Converter pool
+	var wg sync.WaitGroup
+	wg.Add(workerCount)
+	for i := 0; i < workerCount; i++ {
+		go func() {
+			defer wg.Done()
+			for sr := range rawCh {
+				// Now it's this block's turn
+				ethBlock := &pbeth.Block{}
+				if err := sr.resp.Block.UnmarshalTo(ethBlock); err != nil {
+					fmt.Printf("failed to unmarshal block (seq: %d): %v\n", sr.seq, err)
+					continue
+				}
+				block, err := convertFirehoseBlockToGethBlock(ethBlock, chainID, externalRpc, endpoint)
+				if err != nil {
+					fmt.Printf("failed to convert block %d: %v\n", ethBlock.Number, err)
+					continue
+				}
+				blockCh <- seqBlock{seq: sr.seq, block: block}
+			}
+		}()
+	}
+
+	// Close blockCh when all workers are done
+	go func() {
+		wg.Wait()
+		close(blockCh)
+	}()
+
+	// Stage 3: Batching and processing in order
+	go func() {
+		var (
+			blocks   []*types.Block
+			batchNum int
+			nextSeq  uint64 = 0
+			buffer          = make(map[uint64]*types.Block)
+		)
+		for sb := range blockCh {
+			buffer[sb.seq] = sb.block
+			// Drain in-order blocks from buffer
+			for {
+				block, ok := buffer[nextSeq]
+				if !ok {
+					break
+				}
+				blocks = append(blocks, block)
+				delete(buffer, nextSeq)
+				nextSeq++
+				if len(blocks) >= batchSize {
+					if err := handler(blocks, batchNum); err != nil {
+						errCh <- err
+						return
+					}
+					*startBlock = int(blocks[len(blocks)-1].NumberU64() + 1)
+					batchNum++
+					blocks = blocks[:0]
+				}
+			}
+		}
+		// Process any remaining blocks
+		if len(blocks) > 0 {
+			if err := handler(blocks, batchNum); err != nil {
+				errCh <- err
+				return
+			}
+			*startBlock = int(blocks[len(blocks)-1].NumberU64() + 1)
+		}
+		close(doneCh)
+	}()
+
+	// Wait for completion or error
+	select {
+	case err := <-errCh:
+		return err
+	case <-doneCh:
+		return nil
+	}
+}
+
+func isRetryableError(err error) bool {
+	errStr := err.Error()
+	// Check for common retryable errors
+	retryableErrors := []string{
+		"unexpected EOF",
+		"connection reset by peer",
+		"broken pipe",
+		"context deadline exceeded",
+		"transport is closing",
+		"code = Unavailable",
+		"code = Internal",
+		"code = DeadlineExceeded",
+		"rpc error",
+	}
+
+	for _, retryableErr := range retryableErrors {
+		if strings.Contains(errStr, retryableErr) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Checking error by string is never good, must be done only when strictly necessary.